### ISSUE TITLE
Verifying batch_grad for a conv-deconv network

### ISSUE THREAD

**Issue by sayantanauddy** [f-dangel/backpack#219]
Thanks for providing this very useful library.

I was trying to use backpack for a tiny network with Conv2d, BatchNorm2d, and ConvTranspose2d layers.

```python
class TinyConvDeconv(torch.nn.Module):
    def __init__(self, use_bn=True):       
        super(TinyConvDeconv, self).__init__()
        self.use_bn = use_bn
        self.conv1=nn.Conv2d(in_channels=3,out_channels=16, kernel_size=4)
        nn.init.xavier_uniform_(self.conv1.weight) #Xaviers Initialisation
        if self.use_bn:
            self.bn1 = nn.BatchNorm2d(16)
        self.swish1= nn.ReLU()
        self.deconv1=nn.ConvTranspose2d(in_channels=16,out_channels=3,kernel_size=4)
        nn.init.xavier_uniform_(self.deconv1.weight)
        self.swish4=nn.ReLU()

    def forward(self,x):
        out=self.conv1(x)
        out=self.swish1(self.bn1(out) if s
[backpack_conv_deconv.zip](https://github.com/f-dangel/backpack/files/7155414/backpack_conv_deconv.zip)
[backpack_conv_deconv.zip](https://github.com/f-dangel/backpack/files/7155432/backpack_conv_deconv.zip)
elf.use_bn else out)
        out=self.deconv1(out)
        out=self.swish4(out)
        return(out)
```
I set the network mode to eval and then tried to replicate the example [here](https://docs.backpack.pt/en/master/use_cases/example_first_order_resnet.html#sphx-glr-use-cases-example-first-order-resnet-py), where the per-sample gradients are verified. Although I was able to replicate the above example (using the provided ResNet), I could not do the same for my tiny conv-deconv network.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

from torchsummary import summary

from backpack import backpack, extend
from backpack.extensions import BatchGrad

import warnings
# To remove backpack warning about using a non-full backward hook
warnings.filterwarnings('ignore')

BATCH_SIZE = 32
IMAGE_SIZE = 64
torch.manual_seed(0)
[backpack_conv_deconv.zip](https://github.com/f-dangel/backpack/files/7155438/backpack_conv_deconv.zip)

DEVICE = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

# Use random tensors as data
pseudo_x = torch.rand((BATCH_SIZE, 3, IMAGE_SIZE, IMAGE_SIZE)).to(DEVICE)
pseudo_y = torch.rand((BATCH_SIZE, 3, IMAGE_SIZE, IMAGE_SIZE)).to(DEVICE)

conv_deconv = TinyConvDeconv().to(DEVICE)
conv_deconv.eval()

conv_deconv = extend(conv_deconv)

# At this stage BN stats are fixed to 0 (mean) and 1 (var)
print(f'Network train set? {conv_deconv.training}')

conv_deconv.zero_grad()
loss = F.mse_loss(conv_deconv(pseudo_x), pseudo_y, reduction="mean")

with backpack(BatchGrad()):
    loss.backward()

print("{:<20}  {:<40} {:<20}".format("Param", "grad", "grad (batch)"))
print("-" * 100)
for name, p in conv_deconv.named_parameters():
    if (not 'bn' in name):
        print(f'{name:<20}, {str(p.grad.shape):<40}, {str(p.grad_batch.shape):<20}')

sample_to_check = 1

x_to_check = pseudo_x[sample_to_check, :].unsqueeze(0)
y_to_check = pseudo_y[sample_to_check].unsqueeze(0)

conv_deconv.zero_grad()
loss = F.mse_loss(conv_deconv(x_to_check), y_to_check)
loss.backward()

print("Do the individual gradients match?")
for name, p in conv_deconv.named_parameters():
    if (not 'bn' in name):
        match = torch.allclose(p.grad_batch[sample_to_check, :], p.grad, atol=1e-5)
        print("{:<20} {}".format(name, match))
```

I used the same steps as the [example](https://docs.backpack.pt/en/master/use_cases/example_first_order_resnet.html#sphx-glr-use-cases-example-first-order-resnet-py), but I could not figure out why the gradients computed for a single sample do not match the grad_batch computed by backpack. Perhaps I am missing something?

The attached notebook also has the code shown here. I use torch 1.9.0+cu102. Any help is appreciated.

**Comment by f-dangel** [comment]
Hi, and a big thanks for the detailed description of your problem.

The two gradients you compare are off by a constant scaling factor introduced by the `MSELoss`:
- In the forward pass with the full mini-batch, `reduction="mean"` divides the sum of squared residuals by a factor of `BATCH_SIZE * 3 * IMAGE_SIZE * IMAGE_SIZE`. This factor carries over to the individual gradients, as described in the documentation of [`BatchGrad`](https://docs.backpack.pt/en/master/extensions.html#backpack.extensions.BatchGrad).
- In the forward pass with one sample, `reduction="mean"` divides the sum of squared residuals by a factor of `3 * IMAGE_SIZE * IMAGE_SIZE`.

As a result, in your example `p.grad_batch[sample_to_check]` is the same as `p.grad / BATCH_SIZE`.

That said,  **your example will work with one of these modifications**:
1. **Turn off reduction factors:** Replace both `F.mse_loss(...)` by `F.mse_loss(..., reduction="sum")`
2. **Manually correct the scaling:** Keep `reduction="mean"`, but replace
    ```python
    match = torch.allclose(p.grad_batch[sample_to_check, :], p.grad, atol=1e-5)
    ```
    with
    ```python
    match = torch.allclose(p.grad_batch[sample_to_check, :], p.grad / BATCH_SIZE, atol=1e-5)
    ```

Please let us know if this solves your issue, and if you have any suggestions how to improve the documentation.

Best

**Comment by sayantanauddy** [comment]
Thanks so much for the quick answer! This does indeed solve my problem.

Regarding the documentation, I realize that things will improve as the library matures further, but maybe you will consider revising the [section](https://docs.backpack.pt/en/master/good-to-know.html#not-yet-supported-models) about what is currently supported and what is not supported by Backpack, with a bit more clarity on why certain things are not currently supported.

Furthermore, I found the requirement of a standard `forward()` function and the use of `Sequential` to be a bit confusing. This is strictly true only for second-order quantities, right? [This page](https://docs.backpack.pt/en/master/supported-layers.html) mentions that first-order extensions that extract information **might work** outside of this framework, but it is not tested. Under what circumstances may the use of a non-standard `forward()` be a problem for first-order extensions?

Thanks again for your help!

**Comment by f-dangel** [comment]
Thanks for your feedback. We will try to clarify the limitations in future releases.

A general hint: Whenever you are unsure, try double-checking your results with `torch.autograd`. You can find inspiration in some of the [examples & use cases](https://docs.backpack.pt/en/master/use_cases/index.html#) in the documentation, which follow such a presentation style.
